File: python/src/application.py
import glob
import inspect
import os
import shutil


class Application:
    """Deletes files in a directory tree matching a glob pattern, with a
    dry-run mode that prints what would be removed without touching the
    filesystem."""

    class InvalidModeError(Exception):
        pass

    @classmethod
    def run(cls, dirname: str = '.', pattern: str = '*',
            mode: str = 'd') -> None:
        instance = cls(dirname=dirname, pattern=pattern, mode=mode)
        instance.validate_mode()
        instance._run()

    def __init__(self, dirname: str = '.', pattern: str = '*',
                 mode: str = 'd') -> None:
        self._dirname = dirname
        self._pattern = pattern
        self._mode = mode
        self._files = glob.glob(
            os.path.join(dirname, '**', pattern), recursive=True)

    def validate_mode(self) -> None:
        match self._mode:
            case 'd' | 'e':
                return
            case _:
                raise self.InvalidModeError(
                    f'{self._mode} is invalid mode. '
                    'Provide either `d`(default) or `e`.'
                )

    # private

    def _run(self) -> None:
        self._output(
            f'Target dirname is {os.path.abspath(self._dirname)}')
        if not self._files:
            self._announce_empty()
            return
        self._announce_start()
        self._clean_files()
        self._announce_finish()

    def _announce_empty(self) -> None:
        self._output(
            f'========== [{self._exec_mode()}] '
            f'No {self._pattern} Remains ==========')

    def _announce_start(self) -> None:
        self._output(
            f'========== [{self._exec_mode()}] '
            f'Total File Count to Clean: {len(self._files)} ==========')
        self._output(
            f'========== [{self._exec_mode()}] '
            f'Start Cleaning {self._pattern} ==========')

    def _clean_files(self) -> None:
        for file in self._files:
            self._output(
                f'========== [{self._exec_mode()}] '
                f'Cleaning {file} ==========')
        if self._mode == 'e':
            for file in self._files:
                if os.path.isdir(file):
                    shutil.rmtree(file, ignore_errors=True)
                else:
                    os.remove(file)

    def _announce_finish(self) -> None:
        self._output(
            f'========== [{self._exec_mode()}] '
            f'Cleaned {self._pattern} ==========')
        self._output(
            f'========== [{self._exec_mode()}] '
            f'Total Cleaned File Count: {len(self._files)} ==========')

    def _exec_mode(self) -> str:
        return 'EXECUTION' if self._mode == 'e' else 'DRY RUN'

    def _test_env(self) -> bool:
        stack = inspect.stack()
        if not stack:
            return False
        return 'pytest' in os.path.basename(stack[-1].filename)

    def _output(self, message: str) -> None:
        if not self._test_env():
            print(message)

File: ruby/src/application.rb
# frozen_string_literal: true
# rbs_inline: enabled

# Deletes files in a directory tree matching a glob pattern, with a dry-run
# mode that prints what would be removed without touching the filesystem.
class Application
  class InvalidModeError < StandardError; end

  # @rbs dirname: String
  # @rbs pattern: String
  # @rbs mode: String
  # @rbs return: void
  def self.run(dirname: '.', pattern: '*', mode: 'd')
    instance = new(dirname:, pattern:, mode:)
    instance.validate_mode!
    instance.run
  end

  # @rbs dirname: String
  # @rbs pattern: String
  # @rbs mode: String
  # @rbs return: void
  def initialize(dirname: '.', pattern: '*', mode: 'd')
    @dirname = dirname
    @pattern = pattern
    @mode    = mode
    @files   = Dir.glob(File.join(dirname, '**', pattern))
  end

  # @rbs return: void
  def validate_mode!
    case mode
    when 'd', 'e'
      nil
    else
      raise InvalidModeError, "#{mode} is invalid mode. Provide either `d`(default) or `e`."
    end
  end

  # @rbs return: void
  def run
    output "Target dirname is #{File.absolute_path(dirname)}"
    return announce_empty if files.empty?

    announce_start
    clean_files
    announce_finish
  end

  private

  attr_reader :dirname, :pattern, :mode, :files

  # @rbs return: void
  def announce_empty
    output "========== [#{exec_mode}] No #{pattern} Remains =========="
  end

  # @rbs return: void
  def announce_start
    output "========== [#{exec_mode}] Total File Count to Clean: #{files.length} =========="
    output "========== [#{exec_mode}] Start Cleaning #{pattern} =========="
  end

  # @rbs return: void
  def clean_files
    files.each { |file| output "========== [#{exec_mode}] Cleaning #{file} ==========" }
    FileUtils.rm_rf(files) if mode == 'e'
  end

  # @rbs return: void
  def announce_finish
    output "========== [#{exec_mode}] Cleaned #{pattern} =========="
    output "========== [#{exec_mode}] Total Cleaned File Count: #{files.length} =========="
  end

  # @rbs return: String
  def exec_mode
    @exec_mode ||= mode == 'e' ? 'EXECUTION' : 'DRY RUN'
  end

  # @rbs return: bool
  def test_env?
    caller(0..0).first.split('/').last.include?('minitest.rb')
  end

  # @rbs message: String
  # @rbs return: void
  def output(message)
    puts message unless test_env?
  end
end
